bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableDataSettingsFactory.java

```diff
@@ -0,0 +1,310 @@
+
+  /**
+   * To create an instance of {@link BigtableDataSettings} from {@link BigtableOptions}.
+   *
+   * @param configuration a {@link BigtableOptions} object.
+   * @return a {@link BigtableDataSettings} object.
+   * @throws IOException if any.
+   */
+  public static BigtableDataSettings fromBigtableOptions(final BigtableOptions options)
+      throws IOException, GeneralSecurityException {
+    BigtableDataSettings.Builder builder = BigtableDataSettings.newBuilder();
+
+    InstanceName instanceName = InstanceName.newBuilder().setProject(options.getProjectId())
+        .setInstance(options.getInstanceId()).build();
+    builder.setInstanceName(instanceName);
+    builder.setAppProfileId(options.getAppProfileId());
+
+    LOG.debug("endpoint host %s.", options.getDataHost());
+    LOG.debug("endpoint host %s.", options.getPort());
+    builder.setEndpoint(options.getDataHost() + ":" + options.getPort());
+
+    buildCredentialProvider(builder, options.getCredentialOptions());
+
+    buildBulkOptions(builder, options);
+
+    buildCheckAndMutateRow(builder, options);
+
+    buildReadModifyWrite(builder, options);
+
+    buildReadRows(builder, options);
+
+    buildMutateRow(builder, options);
+
+    buildSampleRowKeys(builder, options);
+
+    // TODO: would it map to GrpcHeaderInterceptor? or we should build userAgent
+    // using ManagedChannelBuilder AND
+    builder.setHeaderProvider(
+      FixedHeaderProvider.create(GrpcUtil.USER_AGENT_KEY.name(), options.getUserAgent()));
+
+    // TODO: implementation for channelCount or channelPerCPU
+    ManagedChannelBuilder channelBuilder = ManagedChannelBuilder
+        .forAddress(options.getDataHost(), options.getPort())//
+        .userAgent(options.getUserAgent());
+
+    if (options.usePlaintextNegotiation()) {
+      channelBuilder.usePlaintext();
+    }
+    builder.setTransportChannelProvider(
+      FixedTransportChannelProvider.create(GrpcTransportChannel.create(channelBuilder.build())));
+
+    return builder.build();
+  }
+
+  /**
+   * This method is use to build BatchSettings.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param options a {@link BigtableOptions} object.
+   */
+  private static void buildBulkOptions(Builder builder, BigtableOptions options) {
+    BulkOptions bulkOptions = options.getBulkOptions();
+    BatchingSettings.Builder batchSettingsBuilder = BatchingSettings.newBuilder();
+
+    // NOTE: FlowControlSettings by default uses LimitExceededBehavior.Block
+    FlowControlSettings.Builder flowControlBuilder =
+        FlowControlSettings.newBuilder()
+            .setMaxOutstandingRequestBytes(bulkOptions.getMaxMemory());
+
+    long autoFlushMs = bulkOptions.getAutoflushMs();
+    if (autoFlushMs > 0) {
+      batchSettingsBuilder.setDelayThreshold(Duration.ofMillis(autoFlushMs));
+    }
+    long maxIn = bulkOptions.getMaxInflightRpcs();
+    if(maxIn > 0) {
+      flowControlBuilder.setMaxOutstandingElementCount(maxIn);
```

Comment:
This should be maxInflightRpcs * bulkMaxRowKeyCount

Reply:
oh, will update the same.